### PR TITLE
Add verify and verify_m functions

### PR DESCRIPTION
**Pull request by elliotwutingfeng**
## Changes

- Added **verify()** and **verify_m()** functions to verify passwords against hashes, similar to the **verify()** method in the _argon2-cffi_ library.

### CODE REVIEW

File: balloon.py [hunk 1]
```diff
@@ -238,3 +239,56 @@ def balloon_m_hash(password: str, salt: str) -> str:
     return balloon_m(
         password, salt, space_cost, time_cost, parallel_cost, delta=delta
     ).hex()
+
+
+def verify(
+    password: str, salt: str, space_cost: int, time_cost: int, delta: int, hash: str
```
Comment (daxpedda) [inline balloon.py:245]: Can we put `hash` at the start?
We can then also make `delta` 3 by default again.

File: balloon.py [hunk 2]
```diff
@@ -238,3 +239,56 @@ def balloon_m_hash(password: str, salt: str) -> str:
...
+
+
+def verify_m(
+    password: str,
+    salt: str,
+    space_cost: int,
+    time_cost: int,
+    parallel_cost: int,
+    delta: int,
+    hash: str,
```
Comment (daxpedda) [inline balloon.py:273]: Same as above.

File: test_vectors.py [hunk 3]
```diff
@@ -43,19 +44,18 @@ def test_vectors(self):
...
+            test_params = list(test_vector.values())
+            self.assertEqual(balloon(*test_params[:4]).hex(), test_vector["output"])
             self.assertEqual(
                 balloon_hash(test_vector["password"], test_vector["salt"]),
                 balloon(test_vector["password"], test_vector["salt"], 16, 20, 4).hex(),
             )
+            self.assertTrue(verify(*test_params[:4], 3, test_vector["output"]))
+            self.assertFalse(
+                verify(
+                    *test_params[:4], 3, "%064x" % (int(test_vector["output"], 16) - 1)
```
Comment (daxpedda) [inline test_vectors.py:56]:
```py
"%064x" % (int(test_vector["output"], 16) - 1
```
What is that?

**Review by daxpedda — CHANGES_REQUESTED** [review]
Otherwise great.

File: test_vectors.py (hunk 3 above)
Comment (elliotwutingfeng) [reply test_vectors.py:56]: It converts hash to int, subtracts 1 and formats it back as a 64-byte hex string. Intent is to assert that verify() will return False if the provided sha256 hash does not match.

Comment (daxpedda) [reply test_vectors.py]: Interesting stuff! Can we just make a 64 byte array filled with nulls instead?

Comment (elliotwutingfeng) [reply test_vectors.py]: That works too. Just looked up this, https://www.quora.com/What-data-produces-a-SHA256-hash-of-all-zero-bits. So I think it should be safe.

File: test_vectors.py [hunk 4]
```diff
@@ -43,19 +44,14 @@ def test_vectors(self):
...
-                test_vector["output"],
-            )
+            test_params = list(test_vector.values())
+            self.assertEqual(balloon(*test_params[:4]).hex(), test_vector["output"])
             self.assertEqual(
                 balloon_hash(test_vector["password"], test_vector["salt"]),
                 balloon(test_vector["password"], test_vector["salt"], 16, 20, 4).hex(),
             )
+            self.assertTrue(verify(test_vector["output"], *test_params[:4]))
+            self.assertFalse(verify("0" * 64, *test_params[:4]))
```
Comment (daxpedda) [inline test_vectors.py:54]:
```py
"0" * 64
```
Python keeps shocking me.

**Review by daxpedda — APPROVED** [review]
LGTM, waiting for CI.

**Comment by daxpedda** [comment]
Hm, I guess GitHub actions broke. I will check in later.

**Pull request merged**